crates/consensus/src/transaction/recovered.rs

```diff
@@ -0,0 +1,123 @@
+        (self.signed_transaction, self.signer)
+    }
+
+    /// Create [`Recovered`] from the given transaction and [`Address`] of the signer.
+    ///
+    /// Note: This does not check if the signer is the actual signer of the transaction.
+    #[inline]
+    pub const fn new_unchecked(signed_transaction: T, signer: Address) -> Self {
+        Self { signed_transaction, signer }
+    }
+
+    /// Applies the given closure to the inner transactions.
+    pub fn map_transaction<Tx>(self, f: impl FnOnce(T) -> Tx) -> Recovered<Tx> {
+        Recovered::new_unchecked(f(self.signed_transaction), self.signer)
+    }
+}
+
+impl<T: Encodable> Encodable for Recovered<T> {
+    /// This encodes the transaction _with_ the signature, and an rlp header.
+    ///
+    /// Refer to docs for [`TransactionSigned::encode`] for details on the exact format.
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        self.signed_transaction.encode(out)
+    }
+
+    fn length(&self) -> usize {
+        self.signed_transaction.length()
+    }
+}
+
+impl<T: Decodable + SignerRecoverable> Decodable for Recovered<T> {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let signed_transaction = T::decode(buf)?;
+        let signer = signed_transaction.recover_signer().map_err(|_| {
+            alloy_rlp::Error::Custom("Unable to recover decoded transaction signer.")
+        })?;
+        Ok(Self::new_unchecked(signed_transaction, signer))
+    }
+}
+
+impl<T: Encodable2718> Encodable2718 for Recovered<T> {
+    fn type_flag(&self) -> Option<u8> {
+        self.signed_transaction.type_flag()
+    }
+
+    fn encode_2718_len(&self) -> usize {
+        self.signed_transaction.encode_2718_len()
+    }
+
+    fn encode_2718(&self, out: &mut dyn alloy_rlp::BufMut) {
+        self.signed_transaction.encode_2718(out)
+    }
+
+    fn trie_hash(&self) -> B256 {
+        self.signed_transaction.trie_hash()
+    }
+}
+
+/// A type that can recover the signer of a transaction.
+///
+/// This is a helper trait that only provides the ability to recover the signer (address) of a
+/// transaction.
+///
+/// This
+pub trait SignerRecoverable {
+    /// Recover signer from signature and hash.
+    ///
+    /// Returns an error if the transaction's signature is invalid following [EIP-2](https://eips.ethereum.org/EIPS/eip-2).
+    ///
+    /// Note:
+    ///
+    /// This can fail for some early ethereum mainnet transactions pre EIP-2, use
+    /// [`Self::recover_signer_unchecked`] if you want to recover the signer without ensuring that
+    /// the signature has a low `s` value.
+    fn recover_signer(&self) -> Result<Address, alloy_primitives::SignatureError>;
+
+    /// Recover signer from signature and hash _without ensuring that the signature has a low `s`
+    /// value_.
+    ///
+    /// Returns an error if the transaction's signature is invalid.
+    fn recover_signer_unchecked(&self) -> Result<Address, alloy_primitives::SignatureError>;
```

Comment:
btw was recently thinking about how we could avoid embedding the `_unchecked` stuff into the abstraction

could this be a part of tx validation which is aware of the timestamp and can reject transactions outside of the actual recovery logic?

Reply:
hmm, this could likely be a bit tricky because we also have this a ton in rpc, but would actually be nicer.

let's proceed with unchecked and think how we could phase that out?

Reply:
sg!